### Start with a Business Goal

Hi everyone!

Before you start cleaning a dataset, make sure you understand the **business objective**. Without a clear goal, you’ll likely get stuck in the middle of the work.

---

### No Goal? Use AI to Brainstorm

If you're unsure, ask ChatGPT:

- “Here’s my dataset — what questions would a marketing manager ask?”
- “Based on this job role, what should a data analyst analyze?”

This helps create real business questions.

---

### Project Business Goal

#### Sales & Revenue
- Total sales over time (daily/weekly/monthly)?
- When do we earn the most?
- Top-selling and low-selling items?
- Sales by category (coffee, tea, bakery)?
- Any seasonal or daily patterns?

---

Let’s clean the data with these questions in mind!


### 1.Import necessary libraries

In [3]:
import pandas as pd
import numpy as np

### 2.Load the dataset

In [4]:
df = pd.read_csv('dirty_cafe_sales.csv')

### 3.Explore the dataset
- Identify grain, measures, and dimensions
  
  - Grain: What does each row represent in the dataset

  - Measures: Which columns are numbers or quantitative data

  - Dimensions: Which columns are categories or qualitative data
- Identify Critical vs non-critical columns
  
  - Which columns is high quality (80% complete and accurate)

  - Which columns are less important
- Understand column definitions
  
  - Actual definitions of each of these columns

In [5]:
df.sample(15)       # Get random samples for understand the rows

Transaction ID      Item Quantity Price Per Unit Total Spent  \
5866    TXN_3642052      Cake        1            3.0     UNKNOWN   
347     TXN_7140471      Cake        1            3.0         3.0   
5600    TXN_4908458       Tea        4            1.5         6.0   
8100    TXN_3584252     Salad      NaN            5.0        10.0   
8364    TXN_7914312  Sandwich        5            4.0        20.0   
8922    TXN_9431971  Smoothie        1            4.0         4.0   
2850    TXN_8730549  Smoothie        4            4.0        16.0   
1435    TXN_4901854    Cookie        5            1.0         5.0   
7880    TXN_8156455      Cake        2            3.0         6.0   
9350    TXN_4459351  Smoothie        5            4.0        20.0   
2753    TXN_4624436  Sandwich        1            4.0         4.0   
1336    TXN_6899834      Cake        3            3.0         9.0   
1170    TXN_7626939    Coffee  UNKNOWN            2.0        10.0   
1359    TXN_1388606      Cake        3            3.0         9.0   
4605    TXN_7683607      Cake        1            3.0         3.0   

      Payment Method  Location Transaction Date  
5866             NaN  Takeaway       2023-07-19  
347      Credit Card       NaN       2023-05-25  
5600             NaN       NaN       2023-10-06  
8100  Digital Wallet  Takeaway       2023-12-19  
8364             NaN  In-store       2023-06-06  
8922     Credit Card     ERROR       2023-07-17  
2850             NaN  Takeaway       2023-02-04  
1435           ERROR       NaN       2023-04-23  
7880            Cash       NaN       2023-02-12  
9350     Credit Card       NaN       2023-12-08  
2753             NaN       NaN       2023-05-22  
1336     Credit Card  Takeaway       2023-01-27  
1170             NaN       NaN            ERROR  
1359  Digital Wallet  Takeaway       2023-08-31  
4605            Cash       NaN       2023-03-17

In [6]:
df.shape        # Get how many of rows and columns

(10000, 8)

In [7]:
df.info()       # Get non-null counts and data types of each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [8]:
df.describe()       # Get summary statistics

Transaction ID   Item Quantity Price Per Unit Total Spent  \
count           10000   9667     9862           9821        9827   
unique          10000     10        7              8          19   
top       TXN_1961373  Juice        5            3.0         6.0   
freq                1   1171     2013           2429         979   

        Payment Method  Location Transaction Date  
count             7421      6735             9841  
unique               5         4              367  
top     Digital Wallet  Takeaway          UNKNOWN  
freq              2291      3022              159

In [9]:
df.isna().sum()     # Checking missing values

Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

In [10]:
for col in df.columns:
    print(f'{col}:{df[col].unique()}')      # Get Each column unique values

Transaction ID:['TXN_1961373' 'TXN_4977031' 'TXN_4271903' ... 'TXN_5255387' 'TXN_7695629'
 'TXN_6170729']
Item:['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' nan
 'ERROR' 'Juice' 'Tea']
Quantity:['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN' nan]
Price Per Unit:['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' nan 'ERROR' 'UNKNOWN']
Total Spent:['4.0' '12.0' 'ERROR' '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' nan 'UNKNOWN' '2.0' '1.0' '7.5' '4.5' '1.5']
Payment Method:['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]
Location:['Takeaway' 'In-store' 'UNKNOWN' nan 'ERROR']
Transaction Date:['2023-09-08' '2023-05-16' '2023-07-19' '2023-04-27' '2023-06-11'
 '2023-03-31' '2023-10-06' '2023-10-28' '2023-07-28' '2023-12-31'
 '2023-11-07' 'ERROR' '2023-05-03' '2023-06-01' '2023-03-21' '2023-11-15'
 '2023-06-10' '2023-02-24' '2023-03-25' '2023-01-15' '2023-04-04'
 '2023-03-30' '2023-12-01' '2023-09-18' '2023-06-03' '2023-12-13'
 '2023-04-20' '2023-04-10' '2023-03

In [11]:
for col in df.columns:                                          # Count how many incorrect values present each column
    filter = df[df[col].isin(['ERROR','UNKNOWN'])]
    print(f'{col}:{filter[col].count()}')


Transaction ID:0
Item:636
Quantity:341
Price Per Unit:354
Total Spent:329
Payment Method:599
Location:696
Transaction Date:301


#### For my Exploration
- Each row represnt the sales transtions

- Numeric columns are Quantity, Price Per Unit, and Total Spent

- Categorical columns are Item, Payment Method, and Location

- Time series column are Transaction Date

#### Data Issues Found

**Item**
- 636 wrong entries (`ERROR`, `UNKNOWN`)
- 333 missing values

**Quantity**
- 341 wrong entries
- 138 missing values
- Data type: should be **int**, not string

**Price Per Unit**
- 354 wrong entries
- 179 missing values
- Data type: should be **float**, not string

**Total Spent**
- 329 wrong entries
- 173 missing values
- Data type: should be **float**, not string

**Payment Method**
- 599 wrong entries
- 2,579 missing values

**Location**
- 696 wrong entries
- 3,265 missing values

**Transaction Date**
- 301 wrong entries
- 159 missing values
- Data type: should be **datetime**, not string


#### Solvable Issues

- **Format column names** — keep naming consistent  

- **Standardize string format** — e.g., all lowercase or title case, remove extra spaces

- **Replace incorrect data** (`ERROR`, `UNKNOWN`) with `NaN`  

- **Fix data types** — convert to correct types (int, float, datetime)  

- **Remove duplicates**  

- **Handle missing values** — some columns are related, so we can fill them using each other:  

    - Removing all rows with missing values would reduce the dataset size, which is not good for analysis, so we will impute instead.  

    - **Item** → fill based on `Price Per Unit`  

    - **Quantity** → `Quantity = Total Spent / Price Per Unit`  

    - **Price Per Unit** → `Price Per Unit = Total Spent / Quantity` and fill based on `Item`

    - **Total Spent** → `Total Spent = Price Per Unit * Quantity`  
    
    - **Transaction Date** → remove null values


#### Unsolvable Issues

- **Payment Method** and **Location** have a large number of missing values and incorrect entries.

- There are no related columns to accurately fill these values.  

- Imputing them using mode or prediction could introduce bias.  

- In a real job, we would raise this issue to the team.  

- For this project, we will **exclude these columns** from analysis.



In [12]:
df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='object')

##### Step 1: Remove unwanted columns

In [13]:
df.drop(columns=['Payment Method', 'Location'], axis=1, inplace=True)       # Dropping unwanted columns in the dataset
df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Transaction Date'],
      dtype='object')

##### Step 2: Format column names (consistent style)

In [14]:
df.columns = df.columns.str.lower().str.replace(' ', '_')       # Formatting column names
df.columns

Index(['transaction_id', 'item', 'quantity', 'price_per_unit', 'total_spent',
       'transaction_date'],
      dtype='object')

In [15]:
df

transaction_id      item quantity price_per_unit total_spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
2       TXN_4271903    Cookie        4            1.0       ERROR   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9995    TXN_7672686    Coffee        2            2.0         4.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9997    TXN_5255387    Coffee        4            2.0         8.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

     transaction_date  
0          2023-09-08  
1          2023-05-16  
2          2023-07-19  
3          2023-04-27  
4          2023-06-11  
...               ...  
9995       2023-08-30  
9996       2023-06-02  
9997       2023-03-02  
9998       2023-12-02  
9999       2023-11-07  

[10000 rows x 6 columns]

##### Step 3: Standardize string formatting (lowercase, strip spaces)

In [16]:
for col in df.select_dtypes(include='object'):      # This line of code will help ensure only string type will be formatted.
    df[col] = df[col].str.strip().str.lower()
df

transaction_id      item quantity price_per_unit total_spent  \
0       txn_1961373    coffee        2            2.0         4.0   
1       txn_4977031      cake        4            3.0        12.0   
2       txn_4271903    cookie        4            1.0       error   
3       txn_7034554     salad        2            5.0        10.0   
4       txn_3160411    coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9995    txn_7672686    coffee        2            2.0         4.0   
9996    txn_9659401       NaN        3            NaN         3.0   
9997    txn_5255387    coffee        4            2.0         8.0   
9998    txn_7695629    cookie        3            NaN         3.0   
9999    txn_6170729  sandwich        3            4.0        12.0   

     transaction_date  
0          2023-09-08  
1          2023-05-16  
2          2023-07-19  
3          2023-04-27  
4          2023-06-11  
...               ...  
9995       2023-08-30  
9996       2023-06-02  
9997       2023-03-02  
9998       2023-12-02  
9999       2023-11-07  

[10000 rows x 6 columns]

##### Step 4: Replace incorrect values (`ERROR`, `UNKNOWN`, empty) with `nan`

In [17]:
df.replace(["unknown", "error", "nan", ""], np.nan, inplace=True)        # Replace all wrong entries with NaN.

In [18]:
df.isna().sum()     # You will notice all columns null values are increased because of replacing incorrect values into nulls

transaction_id        0
item                969
quantity            479
price_per_unit      533
total_spent         502
transaction_date    460
dtype: int64

In [19]:
df['quantity'].unique()

array(['2', '4', '5', '3', '1', nan], dtype=object)

In [20]:
df['price_per_unit'].unique()

array(['2.0', '3.0', '1.0', '5.0', '4.0', '1.5', nan], dtype=object)

In [21]:
df['total_spent'].unique()

array(['4.0', '12.0', nan, '10.0', '20.0', '9.0', '16.0', '15.0', '25.0',
       '8.0', '5.0', '3.0', '6.0', '2.0', '1.0', '7.5', '4.5', '1.5'],
      dtype=object)

##### Step 5: Fix data types (int, float, datetime)

In [22]:
df.dtypes

transaction_id      object
item                object
quantity            object
price_per_unit      object
total_spent         object
transaction_date    object
dtype: object

In [23]:
# Convert columns to appropriate data types using pandas nullable types 
# (Int64), which supports missing values (np.nan).
# Using regular int would cause errors if nulls are present.

df = df.astype({
    'quantity': 'Int64',
    'price_per_unit': 'float',
    'total_spent': 'float'
})

In [24]:
# Convert 'transaction_date' column to datetime format.
# Invalid or missing dates will be set to NaT (missing value) using errors='coerce'.

df['transaction_date'] = pd.to_datetime(df['transaction_date'], errors='coerce')

In [25]:
df.dtypes

transaction_id              object
item                        object
quantity                     Int64
price_per_unit             float64
total_spent                float64
transaction_date    datetime64[ns]
dtype: object

##### Step 6: Remove duplicate rows

In [26]:
df.loc[df.duplicated(keep=False)]       # No duplicates in this dataset

Empty DataFrame
Columns: [transaction_id, item, quantity, price_per_unit, total_spent, transaction_date]
Index: []

##### Step 7: Handle missing values (impute where possible, drop where necessary)

In [27]:
df.isna().sum()

transaction_id        0
item                969
quantity            479
price_per_unit      533
total_spent         502
transaction_date    460
dtype: int64

In [28]:
# Create new menu table (Get unique items and price)

df_price = df[['item','price_per_unit']].dropna().drop_duplicates().sort_values(by='price_per_unit').reset_index(drop=True)
df_price = dict(zip(df_price['item'], df_price['price_per_unit']))
df_price

{'cookie': 1.0,
 'tea': 1.5,
 'coffee': 2.0,
 'cake': 3.0,
 'juice': 3.0,
 'smoothie': 4.0,
 'sandwich': 4.0,
 'salad': 5.0}

In [29]:
# Fill the price_per_unit missing values based on item column with mapping concept

print(df['price_per_unit'].isna().sum())
df['price_per_unit'] = df['price_per_unit'].fillna(df['item'].map(df_price))
print(df['price_per_unit'].isna().sum())

533
54


In [30]:
# Fill the price_per_unit column missing based on related column calculations

print(df['price_per_unit'].isna().sum())
df['price_per_unit'] = df['price_per_unit'].fillna(df['total_spent'] / df['quantity'])
print(df['price_per_unit'].isna().sum())

54
6


In [31]:
df[df['price_per_unit'].isna()]

transaction_id item  quantity  price_per_unit  total_spent  \
1761    txn_3611851  NaN         4            <NA>          NaN   
2289    txn_7524977  NaN         4            <NA>          NaN   
3779    txn_7376255  NaN      <NA>            <NA>         25.0   
4152    txn_9646000  NaN         2            <NA>          NaN   
7597    txn_1082717  NaN      <NA>            <NA>          9.0   
9819    txn_1208561  NaN      <NA>            <NA>         20.0   

     transaction_date  
1761       2023-02-09  
2289       2023-12-09  
3779       2023-05-27  
4152       2023-12-14  
7597       2023-12-13  
9819       2023-08-19

In [32]:
df.isna().sum()

transaction_id        0
item                969
quantity            479
price_per_unit        6
total_spent         502
transaction_date    460
dtype: int64

In [33]:
# Create one more menu table, but this time consider possibilities. If multiple items have the same price. Imputation makes bias, so we avoid those items' records.

df_items = df[['item','price_per_unit']].dropna().drop_duplicates()
df_items = df_items.drop_duplicates(subset='price_per_unit', keep=False).sort_values(by='price_per_unit').reset_index(drop=True)
df_items = dict(zip(df_items['price_per_unit'], df_items['item']))

df_items

{np.float64(1.0): 'cookie',
 np.float64(1.5): 'tea',
 np.float64(2.0): 'coffee',
 np.float64(5.0): 'salad'}

In [34]:
df['item'] = df['item'].fillna(df['price_per_unit'].map(df_items))      # Fill the missings from item column based on price_per_unit column data

In [35]:
df.isna().sum()

transaction_id        0
item                480
quantity            479
price_per_unit        6
total_spent         502
transaction_date    460
dtype: int64

In [36]:
condition = (df['quantity'].isna()) & (df['price_per_unit'].isna()) & (df['total_spent'].isna())        # Check if all calculated columns records are missing
print(df[condition].shape[0])
df[condition]

0


Empty DataFrame
Columns: [transaction_id, item, quantity, price_per_unit, total_spent, transaction_date]
Index: []

In [37]:
condition = (df['quantity'].isna()) & (df['price_per_unit'].isna())      # Compare quantity and price_per_unit missing records
print(df[condition].shape[0])
df[condition]

3


transaction_id item  quantity  price_per_unit  total_spent  \
3779    txn_7376255  NaN      <NA>            <NA>         25.0   
7597    txn_1082717  NaN      <NA>            <NA>          9.0   
9819    txn_1208561  NaN      <NA>            <NA>         20.0   

     transaction_date  
3779       2023-05-27  
7597       2023-12-13  
9819       2023-08-19

In [38]:
condition = (df['quantity'].isna()) & (df['total_spent'].isna())        # Compare quantity and total_spent missing records
print(df[condition].shape[0])
df[condition]

20


transaction_id      item  quantity  price_per_unit  total_spent  \
236     txn_8562645     salad      <NA>             5.0          NaN   
278     txn_3229409     juice      <NA>             3.0          NaN   
641     txn_2962976     juice      <NA>             3.0          NaN   
738     txn_8696094  sandwich      <NA>             4.0          NaN   
2796    txn_9188692      cake      <NA>             3.0          NaN   
3203    txn_4565754  smoothie      <NA>             4.0          NaN   
3224    txn_6297232    coffee      <NA>             2.0          NaN   
3401    txn_3251829       tea      <NA>             1.5          NaN   
4257    txn_6470865    coffee      <NA>             2.0          NaN   
5841    txn_5884081    cookie      <NA>             1.0          NaN   
7029    txn_4628338    coffee      <NA>             2.0          NaN   
7297    txn_9944500  smoothie      <NA>             4.0          NaN   
8021    txn_2428781     salad      <NA>             5.0          NaN   
8443    txn_2023651  sandwich      <NA>             4.0          NaN   
8465    txn_9669616    coffee      <NA>             2.0          NaN   
8479    txn_1547245  sandwich      <NA>             4.0          NaN   
8574    txn_2546684     juice      <NA>             3.0          NaN   
8732    txn_4550558    cookie      <NA>             1.0          NaN   
9590    txn_9924732  sandwich      <NA>             4.0          NaN   
9869    txn_1975184    coffee      <NA>             2.0          NaN   

     transaction_date  
236        2023-05-18  
278        2023-04-15  
641        2023-03-17  
738        2023-05-14  
2796       2023-12-01  
3203       2023-10-06  
3224       2023-04-07  
3401       2023-07-25  
4257       2023-09-18  
5841       2023-07-05  
7029       2023-12-25  
7297       2023-01-03  
8021       2023-05-09  
8443       2023-05-25  
8465       2023-06-03  
8479       2023-09-11  
8574       2023-04-08  
8732       2023-08-04  
9590       2023-01-18  
9869       2023-01-15

In [39]:
condition = (df['price_per_unit'].isna()) & (df['total_spent'].isna())        # Compare price_per_unit and total_spent missing records
print(df[condition].shape[0])
df[condition]

3


transaction_id item  quantity  price_per_unit  total_spent  \
1761    txn_3611851  NaN         4            <NA>          NaN   
2289    txn_7524977  NaN         4            <NA>          NaN   
4152    txn_9646000  NaN         2            <NA>          NaN   

     transaction_date  
1761       2023-02-09  
2289       2023-12-09  
4152       2023-12-14

In [40]:
print(df['quantity'].isna().sum())
df['quantity'] = df['quantity'].fillna(df['total_spent'] / df['price_per_unit'])      # Fill the quantity column missing based on related column calculations
print(df['quantity'].isna().sum())

479
23


In [41]:
print(df['total_spent'].isna().sum())
df['total_spent'] = df['total_spent'].fillna(df['quantity'] * df['price_per_unit'])      # Fill the quantity column missing based on related column calculations
print(df['total_spent'].isna().sum())

502
23


In [42]:
df.isna().sum()

transaction_id        0
item                480
quantity             23
price_per_unit        6
total_spent          23
transaction_date    460
dtype: int64

#### Sanity Check

In [43]:
# Before filling missing 'Item' values using 'Price Per Unit',
# we previously ignored prices that appear for multiple items
# (to avoid incorrect matches).
# Now, we ensure that only prices linked to a single unique item are used.

df[df['item'].isna()]['price_per_unit'].unique()

<FloatingArray>
[3.0, 4.0, <NA>]
Length: 3, dtype: Float64

In [44]:
df[df['price_per_unit'].isna()]     # No more chances to fill the balanced missing values

transaction_id item  quantity  price_per_unit  total_spent  \
1761    txn_3611851  NaN         4            <NA>         <NA>   
2289    txn_7524977  NaN         4            <NA>         <NA>   
3779    txn_7376255  NaN      <NA>            <NA>         25.0   
4152    txn_9646000  NaN         2            <NA>         <NA>   
7597    txn_1082717  NaN      <NA>            <NA>          9.0   
9819    txn_1208561  NaN      <NA>            <NA>         20.0   

     transaction_date  
1761       2023-02-09  
2289       2023-12-09  
3779       2023-05-27  
4152       2023-12-14  
7597       2023-12-13  
9819       2023-08-19

In [45]:
df[df['quantity'].isna()]       # No more chances to fill the balanced missing values

transaction_id      item  quantity  price_per_unit  total_spent  \
236     txn_8562645     salad      <NA>             5.0         <NA>   
278     txn_3229409     juice      <NA>             3.0         <NA>   
641     txn_2962976     juice      <NA>             3.0         <NA>   
738     txn_8696094  sandwich      <NA>             4.0         <NA>   
2796    txn_9188692      cake      <NA>             3.0         <NA>   
3203    txn_4565754  smoothie      <NA>             4.0         <NA>   
3224    txn_6297232    coffee      <NA>             2.0         <NA>   
3401    txn_3251829       tea      <NA>             1.5         <NA>   
3779    txn_7376255       NaN      <NA>            <NA>         25.0   
4257    txn_6470865    coffee      <NA>             2.0         <NA>   
5841    txn_5884081    cookie      <NA>             1.0         <NA>   
7029    txn_4628338    coffee      <NA>             2.0         <NA>   
7297    txn_9944500  smoothie      <NA>             4.0         <NA>   
7597    txn_1082717       NaN      <NA>            <NA>          9.0   
8021    txn_2428781     salad      <NA>             5.0         <NA>   
8443    txn_2023651  sandwich      <NA>             4.0         <NA>   
8465    txn_9669616    coffee      <NA>             2.0         <NA>   
8479    txn_1547245  sandwich      <NA>             4.0         <NA>   
8574    txn_2546684     juice      <NA>             3.0         <NA>   
8732    txn_4550558    cookie      <NA>             1.0         <NA>   
9590    txn_9924732  sandwich      <NA>             4.0         <NA>   
9819    txn_1208561       NaN      <NA>            <NA>         20.0   
9869    txn_1975184    coffee      <NA>             2.0         <NA>   

     transaction_date  
236        2023-05-18  
278        2023-04-15  
641        2023-03-17  
738        2023-05-14  
2796       2023-12-01  
3203       2023-10-06  
3224       2023-04-07  
3401       2023-07-25  
3779       2023-05-27  
4257       2023-09-18  
5841       2023-07-05  
7029       2023-12-25  
7297       2023-01-03  
7597       2023-12-13  
8021       2023-05-09  
8443       2023-05-25  
8465       2023-06-03  
8479       2023-09-11  
8574       2023-04-08  
8732       2023-08-04  
9590       2023-01-18  
9819       2023-08-19  
9869       2023-01-15

In [46]:
df[df['total_spent'].isna()]       # No more chances to fill the balanced missing values

transaction_id      item  quantity  price_per_unit  total_spent  \
236     txn_8562645     salad      <NA>             5.0         <NA>   
278     txn_3229409     juice      <NA>             3.0         <NA>   
641     txn_2962976     juice      <NA>             3.0         <NA>   
738     txn_8696094  sandwich      <NA>             4.0         <NA>   
1761    txn_3611851       NaN         4            <NA>         <NA>   
2289    txn_7524977       NaN         4            <NA>         <NA>   
2796    txn_9188692      cake      <NA>             3.0         <NA>   
3203    txn_4565754  smoothie      <NA>             4.0         <NA>   
3224    txn_6297232    coffee      <NA>             2.0         <NA>   
3401    txn_3251829       tea      <NA>             1.5         <NA>   
4152    txn_9646000       NaN         2            <NA>         <NA>   
4257    txn_6470865    coffee      <NA>             2.0         <NA>   
5841    txn_5884081    cookie      <NA>             1.0         <NA>   
7029    txn_4628338    coffee      <NA>             2.0         <NA>   
7297    txn_9944500  smoothie      <NA>             4.0         <NA>   
8021    txn_2428781     salad      <NA>             5.0         <NA>   
8443    txn_2023651  sandwich      <NA>             4.0         <NA>   
8465    txn_9669616    coffee      <NA>             2.0         <NA>   
8479    txn_1547245  sandwich      <NA>             4.0         <NA>   
8574    txn_2546684     juice      <NA>             3.0         <NA>   
8732    txn_4550558    cookie      <NA>             1.0         <NA>   
9590    txn_9924732  sandwich      <NA>             4.0         <NA>   
9869    txn_1975184    coffee      <NA>             2.0         <NA>   

     transaction_date  
236        2023-05-18  
278        2023-04-15  
641        2023-03-17  
738        2023-05-14  
1761       2023-02-09  
2289       2023-12-09  
2796       2023-12-01  
3203       2023-10-06  
3224       2023-04-07  
3401       2023-07-25  
4152       2023-12-14  
4257       2023-09-18  
5841       2023-07-05  
7029       2023-12-25  
7297       2023-01-03  
8021       2023-05-09  
8443       2023-05-25  
8465       2023-06-03  
8479       2023-09-11  
8574       2023-04-08  
8732       2023-08-04  
9590       2023-01-18  
9869       2023-01-15

In [47]:
df.sample(30)       # Checking that our all cleaning steps are correct in each row

transaction_id      item  quantity  price_per_unit  total_spent  \
8594    txn_8627017       tea         4             1.5          6.0   
4210    txn_6399545  sandwich         4             4.0         16.0   
7494    txn_2241162     juice         3             3.0          9.0   
5337    txn_7321479      cake         1             3.0          3.0   
3939    txn_7705748     juice         3             3.0          9.0   
9732    txn_5697414      cake         5             3.0         15.0   
2570    txn_7580441    coffee         1             2.0          2.0   
2776    txn_5157855     salad         1             5.0          5.0   
8314    txn_8082317     juice         3             3.0          9.0   
4200    txn_7139869       tea         1             1.5          1.5   
2623    txn_2432379    coffee         1             2.0          2.0   
744     txn_2004851     salad         5             5.0         25.0   
269     txn_8669422       NaN         4             3.0         12.0   
1111    txn_6881523     juice         5             3.0         15.0   
5995    txn_1745182    cookie         2             1.0          2.0   
9463    txn_2226371      cake         1             3.0          3.0   
8270    txn_8886791       tea         2             1.5          3.0   
628     txn_3179966       tea         2             1.5          3.0   
7930    txn_4755860     juice         5             3.0         15.0   
2584    txn_3499226     juice         1             3.0          3.0   
3347    txn_2300313    coffee         4             2.0          8.0   
9522    txn_6044312      cake         2             3.0          6.0   
9191    txn_4502922       NaN         1             4.0          4.0   
24      txn_9400181  sandwich         5             4.0         20.0   
7357    txn_9658354       tea         3             1.5          4.5   
3882    txn_4992510    cookie         4             1.0          4.0   
3180    txn_3711712     salad         5             5.0         25.0   
1392    txn_7043825     juice         2             3.0          6.0   
1773    txn_7068653    cookie         2             1.0          2.0   
7830    txn_3493771      cake         2             3.0          6.0   

     transaction_date  
8594       2023-01-28  
4210       2023-03-22  
7494       2023-08-09  
5337       2023-10-31  
3939       2023-07-14  
9732       2023-02-16  
2570       2023-08-09  
2776       2023-08-18  
8314       2023-08-18  
4200       2023-06-22  
2623       2023-10-24  
744               NaT  
269        2023-10-15  
1111       2023-02-24  
5995       2023-09-26  
9463       2023-11-18  
8270       2023-07-09  
628        2023-05-27  
7930       2023-10-28  
2584       2023-09-08  
3347       2023-06-02  
9522       2023-01-03  
9191       2023-06-23  
24         2023-06-03  
7357       2023-11-12  
3882       2023-01-26  
3180       2023-08-20  
1392       2023-02-07  
1773       2023-06-05  
7830       2023-02-13

In [48]:
# Drop all rows with missing values, since there are no further options for imputation.
# Yes, we lose some percentage of data, but thanks to the previous steps, without the above operations, we lose more compared to this approach.

df = df.dropna(subset=list(df.columns))

In [49]:
df.isna().sum()

transaction_id      0
item                0
quantity            0
price_per_unit      0
total_spent         0
transaction_date    0
dtype: int64

In [50]:
df.shape

(9064, 6)

#### Add more column for exploration

- Enhance the **transaction_date** column.

- Extract months(number/text), days, dates, and years for more exploration analysis.



In [51]:
df['transaction_month'] = df['transaction_date'].dt.month       # Extract months in numbers
df

transaction_id      item  quantity  price_per_unit  total_spent  \
0       txn_1961373    coffee         2             2.0          4.0   
1       txn_4977031      cake         4             3.0         12.0   
2       txn_4271903    cookie         4             1.0          4.0   
3       txn_7034554     salad         2             5.0         10.0   
4       txn_3160411    coffee         2             2.0          4.0   
...             ...       ...       ...             ...          ...   
9995    txn_7672686    coffee         2             2.0          4.0   
9996    txn_9659401    cookie         3             1.0          3.0   
9997    txn_5255387    coffee         4             2.0          8.0   
9998    txn_7695629    cookie         3             1.0          3.0   
9999    txn_6170729  sandwich         3             4.0         12.0   

     transaction_date  transaction_month  
0          2023-09-08                  9  
1          2023-05-16                  5  
2          2023-07-19                  7  
3          2023-04-27                  4  
4          2023-06-11                  6  
...               ...                ...  
9995       2023-08-30                  8  
9996       2023-06-02                  6  
9997       2023-03-02                  3  
9998       2023-12-02                 12  
9999       2023-11-07                 11  

[9064 rows x 7 columns]

In [52]:
df['transaction_month_name'] = df['transaction_date'].dt.strftime('%B')      # Extract months in name
df

transaction_id      item  quantity  price_per_unit  total_spent  \
0       txn_1961373    coffee         2             2.0          4.0   
1       txn_4977031      cake         4             3.0         12.0   
2       txn_4271903    cookie         4             1.0          4.0   
3       txn_7034554     salad         2             5.0         10.0   
4       txn_3160411    coffee         2             2.0          4.0   
...             ...       ...       ...             ...          ...   
9995    txn_7672686    coffee         2             2.0          4.0   
9996    txn_9659401    cookie         3             1.0          3.0   
9997    txn_5255387    coffee         4             2.0          8.0   
9998    txn_7695629    cookie         3             1.0          3.0   
9999    txn_6170729  sandwich         3             4.0         12.0   

     transaction_date  transaction_month transaction_month_name  
0          2023-09-08                  9              September  
1          2023-05-16                  5                    May  
2          2023-07-19                  7                   July  
3          2023-04-27                  4                  April  
4          2023-06-11                  6                   June  
...               ...                ...                    ...  
9995       2023-08-30                  8                 August  
9996       2023-06-02                  6                   June  
9997       2023-03-02                  3                  March  
9998       2023-12-02                 12               December  
9999       2023-11-07                 11               November  

[9064 rows x 8 columns]

In [53]:
df['transaction_day_name'] = df['transaction_date'].dt.strftime('%A')      # Extract days in name
df

transaction_id      item  quantity  price_per_unit  total_spent  \
0       txn_1961373    coffee         2             2.0          4.0   
1       txn_4977031      cake         4             3.0         12.0   
2       txn_4271903    cookie         4             1.0          4.0   
3       txn_7034554     salad         2             5.0         10.0   
4       txn_3160411    coffee         2             2.0          4.0   
...             ...       ...       ...             ...          ...   
9995    txn_7672686    coffee         2             2.0          4.0   
9996    txn_9659401    cookie         3             1.0          3.0   
9997    txn_5255387    coffee         4             2.0          8.0   
9998    txn_7695629    cookie         3             1.0          3.0   
9999    txn_6170729  sandwich         3             4.0         12.0   

     transaction_date  transaction_month transaction_month_name  \
0          2023-09-08                  9              September   
1          2023-05-16                  5                    May   
2          2023-07-19                  7                   July   
3          2023-04-27                  4                  April   
4          2023-06-11                  6                   June   
...               ...                ...                    ...   
9995       2023-08-30                  8                 August   
9996       2023-06-02                  6                   June   
9997       2023-03-02                  3                  March   
9998       2023-12-02                 12               December   
9999       2023-11-07                 11               November   

     transaction_day_name  
0                  Friday  
1                 Tuesday  
2               Wednesday  
3                Thursday  
4                  Sunday  
...                   ...  
9995            Wednesday  
9996               Friday  
9997             Thursday  
9998             Saturday  
9999              Tuesday  

[9064 rows x 9 columns]

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9064 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   transaction_id          9064 non-null   object        
 1   item                    9064 non-null   object        
 2   quantity                9064 non-null   Int64         
 3   price_per_unit          9064 non-null   Float64       
 4   total_spent             9064 non-null   Float64       
 5   transaction_date        9064 non-null   datetime64[ns]
 6   transaction_month       9064 non-null   int32         
 7   transaction_month_name  9064 non-null   object        
 8   transaction_day_name    9064 non-null   object        
dtypes: Float64(2), Int64(1), datetime64[ns](1), int32(1), object(4)
memory usage: 699.3+ KB


In [55]:
df.sample(20)

transaction_id      item  quantity  price_per_unit  total_spent  \
8728    txn_6657245  sandwich         1             4.0          4.0   
5180    txn_6534833     salad         4             5.0         20.0   
8887    txn_1041891    cookie         5             1.0          5.0   
5444    txn_7732813       tea         2             1.5          3.0   
7284    txn_8942829     salad         5             5.0         25.0   
9517    txn_3455103     salad         1             5.0          5.0   
862     txn_6439579       tea         5             1.5          7.5   
879     txn_1992966  smoothie         5             4.0         20.0   
7843    txn_9281939       tea         3             1.5          4.5   
8767    txn_5433077      cake         4             3.0         12.0   
3463    txn_3174488    cookie         4             1.0          4.0   
6256    txn_8235381    cookie         3             1.0          3.0   
3244    txn_9202272  sandwich         3             4.0         12.0   
4696    txn_4966149     juice         2             3.0          6.0   
6445    txn_5921766  smoothie         3             4.0         12.0   
5332    txn_7414587       tea         3             1.5          4.5   
1410    txn_5469063    coffee         2             2.0          4.0   
9740    txn_8162690     salad         2             5.0         10.0   
7669    txn_3456748     juice         2             3.0          6.0   
8729    txn_1786141  smoothie         4             4.0         16.0   

     transaction_date  transaction_month transaction_month_name  \
8728       2023-10-25                 10                October   
5180       2023-12-14                 12               December   
8887       2023-08-09                  8                 August   
5444       2023-07-16                  7                   July   
7284       2023-03-11                  3                  March   
9517       2023-04-22                  4                  April   
862        2023-03-23                  3                  March   
879        2023-08-09                  8                 August   
7843       2023-05-22                  5                    May   
8767       2023-11-29                 11               November   
3463       2023-06-28                  6                   June   
6256       2023-03-01                  3                  March   
3244       2023-11-29                 11               November   
4696       2023-04-10                  4                  April   
6445       2023-11-06                 11               November   
5332       2023-03-22                  3                  March   
1410       2023-02-24                  2               February   
9740       2023-05-08                  5                    May   
7669       2023-01-30                  1                January   
8729       2023-03-12                  3                  March   

     transaction_day_name  
8728            Wednesday  
5180             Thursday  
8887            Wednesday  
5444               Sunday  
7284             Saturday  
9517             Saturday  
862              Thursday  
879             Wednesday  
7843               Monday  
8767            Wednesday  
3463            Wednesday  
6256            Wednesday  
3244            Wednesday  
4696               Monday  
6445               Monday  
5332            Wednesday  
1410               Friday  
9740               Monday  
7669               Monday  
8729               Sunday

In [56]:
df.shape

(9064, 9)